### PR TITLE
[WB-8329] Detect if script is run with `python -m ...`

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-8329
Fixes WB-8470

Description
-----------
In this PR: 
- Detect if script is run with `python -m ...` and store that in `wandb.run.settings.program`.
- Fix an issue with a functional test that got broken with `sentry-sdk==1.5.5`.

Testing
-------
Added a hacky test to `wandb_settings_test.py` as a PoC and a proper `yea` test.

2022/02/11: Blocked by https://github.com/wandb/yea/pull/18

### CODE REVIEW

**Comment by dmitryduev** [comment]
@raubitsj this seems to address the issue, but is a bit hacky + I'd appreciate your input on testing this (which is super hacky, probably should be a yea test instead -- would appreciate help on that, if that's the case).

File: wandb/sdk/wandb_settings.py [hunk 1]
```diff
@@ -116,7 +116,10 @@ def _get_program() -> Optional[Any]:
     try:
         import __main__  # type: ignore
 
-        return __main__.__file__
+        if __main__.__spec__ is None:
+            return __main__.__file__
+        # likely run as `python -m ...`
+        return f"{sys.executable} -m {__main__.__spec__.name}"
```
Comment (dmitryduev) [inline wandb/sdk/wandb_settings.py:122]: Shouldn't this be simply `f"{__main__.__spec__.name}"`?

**Comment by raubitsj** [comment]
> @raubitsj this seems to address the issue, but is a bit hacky + I'd appreciate your input on testing this (which is super hacky, probably should be a yea test instead -- would appreciate help on that, if that's the case).
> 
> Haven't tried this, would it work?:
> 
> ```yaml
> id: 0.settings.python-m
> tag:
>     shard: service
> plugin:
>     - wandb
> command:
>     program: python -m module.lib
> assert:
>     - :wandb:runs_len: 1
>     - :op:contains:
>         - :wandb:runs[0][settings][program]
>         - python -m module.lib
>     - :wandb:runs[0][exitcode]: 0
> ```

if it doesnt, we should make it work. either my adding another option.  I think this is a good use of yea.

File: wandb/sdk/wandb_settings.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk/wandb_settings.py:122]: yeah, i think "-m ... " makes sense to me
that seems most consistent with other cases.

**Review by raubitsj — APPROVED** [review]
Please add a yea test before merging to get that line covered!

File: functional_tests/settings/python-m.yea [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+tag:
+    shard: service
+plugin:
+    - wandb
+command:
+    program: python -m module.lib
+assert:
+    - :wandb:runs_len: 1
+    - :op:contains:
+        - :wandb:runs[0][settings][program]
```
Comment (raubitsj) [inline functional_tests/settings/python-m.yea:11]: im confused where did this come from?   settings isnt really known in the backend.

Shouldnt this be:
```
:wandb:runs[0][program]: -m module.lib
```

```

File: functional_tests/settings/module/lib.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import wandb
+
+
+if __name__ == "__main__":
+    run = wandb.init(mode="offline")
```
Comment (raubitsj) [inline functional_tests/settings/module/lib.py:5]: and how does this work.  offline testing?

File: functional_tests/settings/python-m.yea (hunk 2 above)
Comment (raubitsj) [reply functional_tests/settings/python-m.yea:11]: this should run in the standard shard... we need to shard those tests, but your code isnt dependent on wandb-service

**Review by raubitsj — CHANGES_REQUESTED** [review]

File: functional_tests/debug/03-sentry-userproc.py [hunk 4]
```diff
@@ -20,5 +20,7 @@
 
 import wandb
 
-with mock.patch("wandb.sdk.wandb_init._WandbInit.init", side_effect=Exception("injected")) as m:
+
+with mock.patch("wandb.sdk.wandb_init._WandbInit.init", mock.Mock(side_effect=Exception("injected"))):
+    wandb.sdk.wandb_init._WandbInit.init.sentry_repr = None
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply functional_tests/debug/03-sentry-userproc.py:25]: that's a repr func that sentry can then use internally for its reports. by default their repr is too verbose and as i understand, people were complaining that it creates too much noise -- see https://github.com/getsentry/sentry-python/pull/1322

**Review by raubitsj — APPROVED** [review]
Lgtm

**Pull request merged**